File: tests/typesafety/typescript/pipeline_dicts.ts
// expect: passes

function parseOrder(raw: Record<string, unknown>): Record<string, unknown> {
  const quantityText = raw["quantityText"];
  if (typeof quantityText !== "string") {
    throw new Error("quantityText missing");
  }
  return { item: raw["item"], quantity: Number(quantityText) };
}

function price(order: Record<string, unknown>, unitCents: number): Record<string, unknown> {
  const quantity = order["quantity"];
  if (typeof quantity !== "number") {
    throw new Error("quantity missing");
  }
  return { item: order["item"], totalCents: quantity * unitCents };
}

export function quote(raw: Record<string, unknown>): Record<string, unknown> {
  return price(parseOrder(raw), 250);
}

File: tests/typesafety/python/pipeline_dicts.py
# expect: passes
# title: Each step re-checks the dict it was handed
def parse_order(raw: dict[str, object]) -> dict[str, object]:
    quantity_text = raw.get("quantity_text")
    if not isinstance(quantity_text, str):
        raise ValueError("quantity_text missing")
    return {"item": raw.get("item"), "quantity": int(quantity_text)}


def price(order: dict[str, object], unit_cents: int) -> dict[str, object]:
    quantity = order.get("quantity")
    if not isinstance(quantity, int):
        raise ValueError("quantity missing")
    return {"item": order.get("item"), "total_cents": quantity * unit_cents}


def quote(raw: dict[str, object]) -> dict[str, object]:
    return price(parse_order(raw), unit_cents=250)
